### PR TITLE
Scalable go service blog

### PR DESCRIPTION
**Pull request by btalayeminaei**

### CODE REVIEW

File: scalable-go-service/index.md [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+the service if the user has to wait several seconds or even minutes for the
+search results to come back.
+
+## Concurrency and Go
+
+Since we have no influence on how our vendors' APIs are designed, it was clear
+to the team that we had to parallelize our rates requests for individual hotels.
+We are already parallelizing these requests in our services that interact with
+the sunsetting APIs. The existing implementation uses Javascript promises to
+create multiple concurrent requests for hotel rates. We wanted to evaluate this
```
Comment (sdotson) [inline scalable-go-service/index.md]: The shift to present tense feels awkward. Perhaps something like:
```suggestion
We were already parallelizing these requests in our services that interacted with
the sunsetting APIs. The original implementation used Javascript promises to
create multiple concurrent requests for hotel rates. We wanted to evaluate this
```

File: scalable-go-service/index.md [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+   }
+
+   return rates, nil
+}
+
+```
+
+We were quickly able to implement this solution and see an improvement in the
+load time, which was a great proof of concept. However, there were some concerns
+regarding the scalability, configurability, and maintainability, of this
```
Comment (sdotson) [inline scalable-go-service/index.md]:
```suggestion
regarding the scalability, configurability, and maintainability of this
```

File: scalable-go-service/index.md [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+solution. `goroutines` are cheap and a typical `go` program can spawn thousands
+of them during its runtime. However, with this implementation we could
+potentially spawn hundreds of routines at any moment. More concerning is that the
+contents of the request have large consequences on the internal resources of the
+application. Even if thoroughly validated the request to ensure it would not
```
Comment (sdotson) [inline scalable-go-service/index.md]: I think there is some opportunity to smooth out this phrase: `Even if thoroughly validated the request to ensure it would not abuse our internal resources`

File: scalable-go-service/index.md [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+
+The route handler, the token manager, and the rate loader were the three
+components we identified as required for processing a request to load rates. It
+was not immediately obvious how these components would run independently. The
+rate loader takes a validated request from the route handler and needs a token
+before contacting the vendor. However, we noticed that multiple rate loaders can
+execute independently once they have all of the necessary info. Once we
+facilitate proper communication between the components, we can start regarding
+them as workers that receive required parameters from `channels`, perform their
+task, and communicate result through `channels`. Groups of workers that perform
```
Comment (sdotson) [inline scalable-go-service/index.md]:
```suggestion
task, and communicate the result through `channels`. Groups of workers that perform
```

File: scalable-go-service/index.md [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+
+// Stop token loader
+func (tl *TokenLoader) Stop() {
+	close(tl.stop)
+}
+```
+
+## Importance of Communication
+
+As these vendor services grow, the teams has had some discussions around how
```
Comment (sdotson) [inline scalable-go-service/index.md]:
```suggestion
As these vendor services grow, the team has had discussions around how
```

File: scalable-go-service/index.md [hunk 6]
```diff
@@ -0,0 +1,322 @@
+#  Writing Scalable Go Services
+
+## Problem
+
+How do we show the customer a complete set of relevant hotel rates before they
+lose interest and close the tab? Despite trying various strategies to address
+this, the team will likely continue to face this challenge in the future as user
+expectations on speed and accuracy grow. The bottom line is, we need load a lot
```
Comment (andrewgu12) [inline scalable-go-service/index.md]: `we need load` -> `we need to load`

File: scalable-go-service/index.md [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+Since we have no influence on how our vendors' APIs are designed, it was clear
+to the team that we had to parallelize our rates requests for individual hotels.
+We are already parallelizing these requests in our services that interact with
+the sunsetting APIs. The existing implementation uses Javascript promises to
+create multiple concurrent requests for hotel rates. We wanted to evaluate this
+implementation and look for ways to improve it by using a more suitable
+programming language or paradigm that could help the efficiency and design of
+our concurrent program.
+
+If you are not familiar, `go` is a relatively new language and is designed based
```
Comment (andrewgu12) [inline scalable-go-service/index.md]: I think we can change the beginning of this to something about us looking around for different options and eventually settling on go. Currently it just feels like a very awkward transition

File: scalable-go-service/index.md [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+communicating that as we did with other concurrent modules. The failure to
+communicate clearly made this part of the program hard to follow and it would be
+confusing to new engineers. Furthermore, we had moved away from the paradigm
+identified as optimum for the problem. It would be difficult to accurately
+assess the overall solution if we do not consistently follow and communicate the
+strategy.
+
+## Helpful Links
+
+<https://blog.golang.org/share-memory-by-communicating><br/>
```
Comment (rampantmonkey) [inline scalable-go-service/index.md]: This should actually be a linked list, not just different lines.

File: scalable-go-service/index.md [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+## Problem
+
+How do we show the customer a complete set of relevant hotel rates before they
+lose interest and close the tab? Despite trying various strategies to address
+this, the team will likely continue to face this challenge in the future as user
+expectations on speed and accuracy grow. The bottom line is, we need to load a
+lot of rates from many different vendors to show that we have as complete of an
+inventory as any other travel site. And we need to do this fast.
+
+Two of our main hotel vendors have announced major changes to the APIs we use to
```
Comment (rampantmonkey) [inline scalable-go-service/index.md]: This is a better introductory paragraph. It states the explicit motivation that is driving the timing of the change and the value of this work. Then you can use the previous paragraph ("How do we show the customer...") to show the ongoing concerns that drive specific technical challenges of this work.

File: scalable-go-service/index.md [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+The latter case causes a lot of bottleneck for the shopping flow because even
+though the individual requests resolve fairly quickly, the time complexity grows
+linearly with the number of hotels. This would scale poorly as we increase the
+number of hotels we search. It also hinders user experience and confidence in
+the service if the user has to wait several seconds or even minutes for the
+search results to come back.
+
+## Concurrency and Go
+
+Since we have no influence on how our vendors' APIs are designed, it was clear
```
Comment (rampantmonkey) [inline scalable-go-service/index.md]: This paragraph does not provide any new information. Should probably be mostly deleted / merged with the "why go" portion of the next paragraph.

File: scalable-go-service/index.md [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+to the team that we had to parallelize our rates requests for individual hotels.
+We were already parallelizing these requests in our services that interact with
+the sunsetting APIs. The original implementation used Javascript promises to
+create multiple concurrent calls for hotel rates. We wanted to evaluate this
+implementation and look for ways to improve it by using a more suitable
+programming language or paradigm that could help the efficiency and design of
+our concurrent program.
+
+If you are not familiar, `go` is a relatively new language and is designed based
```
Comment (rampantmonkey) [inline scalable-go-service/index.md]: Delete 'if you are not familiar'. The sentence is accurate and useful to a reader regardless of how they feel about their own level of familiarity with go.

File: scalable-go-service/index.md [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+assess the overall solution if we do not consistently follow and communicate the
+strategy.
+
+## Helpful Links
+
+<https://blog.golang.org/share-memory-by-communicating><br/>
+<https://medium.com/@1_00794/parallelism-models-actors-vs-csp-vs-multithreading-f1bab1a2ed6b><br/>
+<https://www.youtube.com/watch?v=cN_DpYBzKso><br/>
+<http://www.usingcsp.com/cspbook.pdf><br/>
+
```
Comment (rampantmonkey) [inline scalable-go-service/index.md:248]: A link to the actual CSP paper would also be great
https://www.cs.cmu.edu/~crary/819-f09/Hoare78.pdf

File: scalable-go-service/index.md [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+statement, which enables interactions with multiple `channels`. So, creating a
+`goroutine` is cheap and we are able to use first-class primitives to facilitate
+conversation between them. But how do we use these features to design and
+implement a scalable service for gathering hotel rates?
+
+## Communicating Sequential Processes
+
+Concurrency in `go` is based on Tony Hoare's CSP, which specifies the patterns
+of interactions between concurrent processes.  Unlike the traditional thread
+model, CSP takes a message passing approach to ensure data integrity rather than
```
Comment (rampantmonkey) [inline scalable-go-service/index.md]: Goroutines and channels have nothing to do with preventing shared access to memory. Mutexes are still used all over the place in go.

Goroutines and channels are the reason why shared, mutable access to memory can happen in the first place.

Comment (rampantmonkey) [reply scalable-go-service/index.md]: Our code happens to use copies of data, and doesn't modify it. But that is a consequence of the choices you made during development, not the primitives themselves.

File: scalable-go-service/index.md [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+locks and mutexes. Go's `goroutines` and `channels` mediate access to shared
+data by passing data references between processes and help structure concurrent
+software. Independent components should run concurrently and communicate with
+other components, if need be. This decouples concurrency from parallelism and
+makes programs easier to digest. Concurrency is the composition of independently
```
Comment (rampantmonkey) [inline scalable-go-service/index.md]: Does explicitly distinguishing between concurrency and parallelism help make your overall point / tell a better story?

File: scalable-go-service/index.md [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+       return rates, fmt.Errorf("no rates found for search criteria")
+   }
+
+   return rates, nil
+}
+
+```
+
+We were quickly able to implement this solution and see an improvement in the
+load time, which was a great proof of concept. However, there were some concerns
```
Comment (rampantmonkey) [inline scalable-go-service/index.md]: Rather than saying that there were concerns, just inline them here. 
> However, this implementation allowed the incoming requests to manipulate resource usage and the number of calls to our external vendors.

File: scalable-go-service/index.md (hunk 8 above)
Comment (rampantmonkey) [reply scalable-go-service/index.md]: Even better, inline these references as links in the main body of the post rather than an appendix at the end.

File: scalable-go-service/index.md [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+		wg.Done()
+	}()
+
+	wg.Add(1)
+	go func() {
+		rateLoaderPool.Start()
+		wg.Done()
+	}()
+
+    // omitting implementation of gracefully shutting down all the workers
```
Comment (rampantmonkey) [inline scalable-go-service/index.md]: you should at least include the `wg.Wait()` line. Otherwise this main function will exit immediately.

File: scalable-go-service/index.md [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+	var rateLoaderPool = RateLoaderPool{
+		Count:         viper.GetInt("rateLoaderCount"),
+		In:            ratesRQChan,
+		TokenRQChan:   tokenRQChan,
+	}
+
+	appRouter := mux.NewRouter()
+	appRouter.HandleFunc("/rates", RatesHandler(ratesRQChan))
+	srv := http.Server{
+		Addr:    viper.GetString("server.host") + ":" + viper.GetString("server.port"),
```
Comment (rampantmonkey) [inline scalable-go-service/index.md]: Hard code the address and count. Or read them directly from environment variables. Introducing a third party package (`viper`) is distracting from the overarching point.

File: scalable-go-service/index.md [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+	    		CheckInDate:             r.CheckInDate,
+	    		CheckOutDate:            r.CheckOutDate,
+	    		RateRSChan:              rateRSChan,
+	    	}
+	    }
+	    for range p.HotelIds {
+	    	res := <-rateRSChan
+            rates = append(rates, res.HotelRateInfo)
+	    }
+		writeResponse(&w, createJsonResponse(r.Context(), rates, http.StatusOK))
```
Comment (rampantmonkey) [inline scalable-go-service/index.md]: `createJsonResponse` is unnecessary in the context of a blog post. just `writeResponse` of `rates`. (It is not exactly how our system works, but it keeps the reader focused)

File: scalable-go-service/index.md [hunk 19]
```diff
@@ -0,0 +1,322 @@
...
+
+```go
+type RateLoaderPool struct {
+	Count         int
+	In            chan RateRQ
+	TokenRQChan   chan<- TokenRQ
+	stop          chan struct{}
+}
+
+// Start initializes and starts n unique rate loader based on the Count parameter.
```
Comment (rampantmonkey) [inline scalable-go-service/index.md]: Comments are weird in a blog post. Either the code is self explanatory or the surrounding paragraphs give enough context.

File: scalable-go-service/index.md [hunk 20]
```diff
@@ -0,0 +1,322 @@
...
+	}
+}
+
+// Stop token loader
+func (tl *TokenLoader) Stop() {
+	close(tl.stop)
+}
+```
+
+## Importance of Communication
```
Comment (rampantmonkey) [inline scalable-go-service/index.md]: This section is confusing. On a meta level it is funny, but I don't think that is the intention.
